XCM Doc Restructure: Covering XCVM (#4579)
w3f/polkadot-wiki: docs/learn/learn-xcvm.md

---

At the core of XCM lies the Cross-Consensus Virtual Machine (XCVM). A “message” in XCM is an XCVM
program, referred to as an **"XCM"** or **"XCMs"** for multiple messages. The XCVM is a
register-based state machine. The state is tracked in domain-specific registers that can hold

> information regarding the temporary state of a particular message. Most of the XCM format comprises

Reviewer: ```suggestion
register-based state machine. The state is tracked in domain-specific registers that hold
information that is used and mutated along the execution of a particular message. Most of the XCM format comprises
```

format comprises these registers and the instructions used to compose XCVM programs.

The XCVM is an ultra-high-level non-Turing-complete computer whose instructions are designed to be
roughly at the same level as transactions in terms of definition. Messages are one or more XCM
instructions executed in order by the XCVM. An XCM is executed until it either runs to the end or
hits an error, at which point it finishes up and halts.

> The primary implementation of the XCVM is the

Reviewer 2: Let's not call it "primary". That would imply we never expect any other implementation to become more significant, which is an overtly political assertion that does not belong in a specification.

"The first implementation" would work just as well.

[`xcm-executor`](https://github.com/paritytech/polkadot/tree/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor).
It follows the XCVM specification provided by Parity. It's engineered to be extendable, providing
maximum customizability when configuring XCM. Because the `xcm-executor` is just an implementation
of XCVM, it's entirely possible to create another implementation if desired.

## XCMs are XCVM Programs

A cross consensus message (XCM) is just a program that runs on the `XCVM`: in other words, one or

> more XCM instructions that are executed by the `xcm-executor`. To learn more about the XCVM and the

Reviewer: ```suggestion
more XCM instructions that are executed by an XCVM implementation, such as the `xcm-executor`. To learn more about the XCVM and the
```

[configure fees](https://github.com/paritytech/polkadot/blob/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor/src/config.rs#L89),
how to
[convert origins](https://github.com/paritytech/polkadot/blob/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor/src/config.rs#L40),
and more.

## Cross Consensus Message (XCM) Anatomy & Flow

An XCM is made up of a list of instructions that are executed in order. There are four different
kinds of XCM instructions:

1. **Instruction** - Results in a state change in the local consensus system or some state change.
2. **Trusted Indication** - Tells the XCVM, or the Executor, that some action has been done before
   already - meaning, this action is now trusted and can be acted on, i.e., in a teleport scenario.
3. **Information** - Provides additional information about a particular origin, usually the result
   of a query, i.e., a `QueryResponse` instruction.
4. **System Notification** - Typically used in the context of when an HRMP channel is being opened,
   closed, or accepted.

Typically, an XCM takes the following path through the XCVM:

> 1.  Instructions within an XCM are read one-by-one by the XCVM. An XCM may old one or more

Reviewer: ```suggestion
1.  Instructions within an XCM are read one-by-one by the XCVM. An XCM may contain one or more
```

    instructions.
2.  The instruction is executed. This means that the current values of the _XCVM registers_, the
    _instruction type_, and the _instruction operands_ are all used to execute some operation, which
    might result in some registers changing their value, or in an error being thrown, which would
    halt execution.
3.  Each subsequent instruction within the XCM is read until the end of the message has been
    reached.

### Example Register: The Holding Register

There are many instructions that depend on the Holding register. The Holding register is an XCVM
register that provides a place for any assets that are in an intermediary state to be held until

> they are deposited to the beneficiary. They require an instruction to place assets within and

Reviewer: ```suggestion
they are deposited to some beneficiary. It requires an instruction to place assets within it and
```

> another to withdraw them. The simplest example of this occurring is the `DepositAsset` instruction,

Reviewer: ```suggestion
another to take them out of it. The simplest example of this is the `DepositAsset` instruction,
```

which in its Rust form looks like this:

```rust
enum Instruction {
    DepositAsset {
        assets: MultiAssetFilter,
        beneficiary: MultiLocation,
    },
    /* snip */
}
```

> This instruction specifies how many assets are going to be withdrawn (useful for fee calculation),
> which assets to filter from the Holding register, and the beneficiary (the recipient of) the assets.

Reviewer: ```suggestion
This instruction specifies which assets (kind and amount), already present in the Holding register, are going to be taken from it and deposited to a beneficiary (recipient).
```

customizability when configuring XCM. Because the `xcm-executor` is just an implementation of XCVM,
it's entirely possible to create another implementation if desired.

## XCMs are XCVM Programs

A cross consensus message (XCM) is just a program that runs on the `XCVM`: in other words, one or
more XCM instructions that are executed by the `xcm-executor`. To learn more about the XCVM and the
XCM format, see the latest
[blog post on XCM v3](https://medium.com/polkadot-network/xcm-part-iii-execution-and-error-management-ceb8155dd166)
by Dr. Gavin Wood.

XCM instructions might change a register, the state of the consensus system, or both. Depending on
the program's goal, whether it is to teleport assets from one chain to another or call a smart
contract on another chain, XCMs usually require changes to the registers before any changes to the
consensus system can be made.

## XCM Executor & Configuration

The XCM Executor's implementation centers around a core piece: the XCM configuration. Each instance
of the Executor must have a valid configuration, which specifies a multitude of options on how a
chain may treat incoming messages via

> [Barriers](https://github.com/paritytech/polkadot/blob/master/xcm/xcm-executor/src/config.rs#L52),

Reviewer 3: Always include permalinks for the GitHub links included in the Wiki

Author: Done!

calculate weight for a message via the
[Weigher](https://github.com/paritytech/polkadot/blob/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor/src/config.rs#L55),
how much weight to purchase via the
[Trader](https://github.com/paritytech/polkadot/blob/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor/src/config.rs#L58),
[configure fees](https://github.com/paritytech/polkadot/blob/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor/src/config.rs#L89),
how to
[convert origins](https://github.com/paritytech/polkadot/blob/26b0c4f6273190f4538b24939a56b6a0b51a344c/xcm/xcm-executor/src/config.rs#L40),

> and more.

Reviewer: We should have a section in the wiki for all the possible configurations. I don't think we have one right now. A short introduction here is the way to go though

Author: Agreed - in the works after the high level stuff is done :D

## Cross Consensus Message (XCM) Anatomy & Flow

An XCM is made up of a list of instructions that are executed in order. There are four different
kinds of XCM instructions:

1. **Instruction** - Results in a state change in the local consensus system or some state change.
2. **Trusted Indication** - Tells the XCVM, or the Executor, that some action has been done before
   already - meaning, this action is now trusted and can be acted on, i.e., in a teleport scenario.
3. **Information** - Provides additional information about a particular origin, usually the result
   of a query, i.e., a `QueryResponse` instruction.
4. **System Notification** - Typically used in the context of when an HRMP channel is being opened,
   closed, or accepted.

> Typically, an XCM takes the following path:

Reviewer: This doesn't look like a series of steps an XCM goes through on the XCVM. It should be something like:
1. An XCM instruction is read
2. The instruction is executed. This means that the current values of the XCVM registers, the instruction type, and the instruction operands are all used to execute some operation, which might result in some registers changing their value, or in an error being thrown, which halts execution
3. The next instruction is read and so on until the end of the message

Author: I suppose it covers more than the XCVM - will modify as needed

1.  The XCM Executor, and XCVM in general, is very much analogous to a CPU. It has various
    registers, which instructions usually modify as needed.
2.  The resulting values from these registers are used later in the transfer. An example of a common
    register used in many instructions is the `Holding` register, which places assets in a holding
    until fulfilled on the destination chain.
3.  The resulting instructions modify the Holding register as needed and execute extrinsics on the
    destination chain to perform the state change. The Executor is able to convert incoming
    instructions into FRAME-compatible origins, which enables the actual state changes to take
    place.

### Example Register: The Holding Register

> There are many instructions that depend on the Holding register. The Holding register is an XCM

Reviewer: ```suggestion
There are many instructions that depend on the Holding register. The Holding register is an XCVM
```

register that provides a place for any assets that are in an intermediary state to be held until

> they are withdrawn to a beneficiary. They require an instruction to place assets within and another

Reviewer: I think using the word "withdraw" for removing assets from the holding register might be confusing since the instructions usually call it `Deposit` and `Withdraw` is the opposite operation

Author: Ah, so to be clear: to _deposit_ means to take from the holding register, while withdraw means to withdraw assets _to_ the holding register?

Reviewer: Yeah, at least that's the logic with the instructions `DepositAsset` and `WithdrawAsset`. They act on accounts, where they `deposit` or `withdraw` but as a side effect end up taking from or putting in the holding register respectively.

to withdraw them. The simplest example of this occurring is the `DepositAsset` instruction, which in
its Rust form looks like this:

```rust
enum Instruction {
    DepositAsset {
        assets: MultiAssetFilter,

>         max_assets: u32,

Reviewer: This operand was removed in v3

        beneficiary: MultiLocation,
    },
    /* snip */
}
```

This instruction specifies how many assets are going to be withdrawn (useful for fee calculation),
which assets to filter from the Holding register, and the beneficiary (the recipient of) the assets.
It is very common for instructions to remove and place assets into the Holding register when
transacting between chains.

### Example: TransferAsset

> An example below illustrates how a chain may transfer assets locally using an XCM. In this message,

Reviewer: It may also transfer assets remotely. I mean, "locally" on a remote chain. I think the wiki calls it remote transfers

}
```

- A `MultiAsset` is a general identifier for an asset. It may represent both fungible and
  non-fungible assets, and in the case of a fungible asset, it represents some defined amount of the
  asset.

- A `MultiLocation` is a relative identifier, meaning that it can only be used to define the the
  relative path between two locations, and cannot generally be used to refer to a location
  universally.

> `TransferAsset` is one of the many instructions that can be executed within an XCM. For more

Reviewer: ```suggestion
`TransferAsset` is one of the many instructions that can be executed within an XCVM. For more
```
or
```suggestion
`TransferAsset` is one of the many instructions that can be contained within an XCM. For more
```

information, please read [XCM Instructions in the wiki](./learn-xcm-instructions.md).

## Locations in XCM

> XCM's abstract nature allows for locations to be relatively abstract notions that point to where but
> also _to who_ a particular action may affect. The `MulitLocation` type is what XCM uses to define
> these locations. A `MultiLocation` is a relative identifier that defines a **relative** path into
> some state-bearing consensus system.

Reviewer: ```suggestion
XCM's general nature requires representing a wide array of "locations" (think relay chains, para chains, solo chains, smart contracts, etc). It also requires specifying accounts or "bodies" inside these systems.
The `MulitLocation` type is what XCM uses to define these locations.
A `MultiLocation` is an identifier that defines a **relative** path into some state-bearing consensus system.
```

It is used to define the relative path between two locations, and cannot generally be used to refer
to a location universally. It is very much akin to how a **relative** filesystem path works and is
dependent on the context on which consensus system the message is being sent from, and what
consensus system it is being sent to be executed.

![XCM MultiLocation](../assets/cross-consensus/multilocation.png)

`MultiLocation` has two primary fields:

- A series of paths, called `Junctions`, which define an interior portion of state to descend into
  it (sometimes called a "sub-consensus" system, such as a smart contract or pallet). An interior
  location may also be used to refer to a Junction, used in the context of "a parachain is an
  interior location of the relay chain", or how a UTXO is interior to Bitcoin's consensus.
- The number of parent junctions at the beginning of a `MultiLocation`'s formation - in other words,
  the number of parent consensus systems above it.

There are a number of various `Junction` variants that may be used to describe a particular
location - whether it's a 32 byte account, a Substrate pallet, or a pluralistic body.

> 

Reviewer: I would also add an example of a multilocation, e. g:

If you were on Parachain X on Polkadot and wanted to reference an account on Statemint (para chain id 1000) you would write the following `MultiLocation`:

```
../Parachain(1000)/AccountId32(<some_id>)
```
which in rust would be:

```rust
MultiLocation {
  parents: 1,
  interior: X2(Parachain(1000), <some_id>.into())
}
```

As illustrated by the `../` or `parents: 1` in Rust, this `MultiLocation` will go "up" by one consensus system and then go "down" the two junctions `Parachain` and `AccountId32`.
Keep in mind this location is relative, if this `MultiLocation` was interpreted by a Polkadot para chain, it would go "up" to the relay chain and then "down" to statemint (para chain id 1000) and to an account with some id.
If it was interpreted in Kusama, the underlying location it identifies would be different.
Some consensus systems might not even be able to interpret this location (Ethereum for example).

Author: This is great, I'll encapsulate this 🎉

}
```

- A `MultiAsset` is a general identifier for an asset. It may represent both fungible and
  non-fungible assets, and in the case of a fungible asset, it represents some defined amount of the
  asset.

- A `MultiLocation` is a relative identifier, meaning that it can only be used to define the the
  relative path between two locations, and cannot generally be used to refer to a location
  universally.

`TransferAsset` is one of the many instructions that can be contained within an XCM. For more
information, please read [XCM Instructions in the wiki](./learn-xcm-instructions.md).

> ## Locations in XCM

Reviewer 2: It might be prudent to mention the Universal Location in its own subsection here.

Author: @gavofyork Added - I'm not entirely sure it's accurate, so let me know if anything needs to be changed! If it's all well, I'll add a graphic to demonstrate a scenario.

As a sidenote, this concept will really come in handy when explaining how a bridge between global networks could actually look like 👍

  non-fungible assets, and in the case of a fungible asset, it represents some defined amount of the
  asset.

- A `MultiLocation` is a relative identifier, meaning that it can only be used to define the the
  relative path between two locations, and cannot generally be used to refer to a location
  universally.

`TransferAsset` is one of the many instructions that can be contained within an XCM. For more
information, please read [XCM Instructions in the wiki](./learn-xcm-instructions.md).

## Locations in XCM

XCM's generic nature involves specifying a wide array of "locations", or any body that is governed
by consensus (parachains, solochains, smart contracts, accounts, etc). These are relatively abstract
notions that point to _where_ but also _to who_ a particular action may affect. The `MulitLocation`
type is what XCM uses to define these locations.

A `MultiLocation` is a relative identifier that defines a **relative** path into some state-bearing
consensus system.

It is used to define the relative path between two locations, and cannot generally be used to refer
to a location universally. It is very much akin to how a **relative** filesystem path works and is

> dependent on the context on which consensus system the message is being sent from, and what
> consensus system it is being sent to be executed.

Reviewer 2: ```suggestion
dependent on the which consensus system the location expression is being evaluated.
```

if this location was defined on another consensus system, such as Kusama. On other consensus
systems, such as Ethereum, it won't be able to interpret it.

### UniversalLocation in XCM

> While a `MultiLocation` may handle relative paths in _local_ consensus, a `UniversalLocation` refers

Reviewer 2: No - a `MultiLocation` is just a relative location. It has no concept of local consensus and can trivially refer to systems outside of local consensus.

> to any global consensus system. A global consensus system is an entity like another relay chain,

Reviewer 2: "...is an entity which provides its own top-level consensus through some non-derivative consensus algorithm able to exist without reference to any other singleton data system".

> Bitcoin, or Ethereum. A body of global governance capable of externalizing its consensus can be

Reviewer 2: No idea what "a body of global governance" means here.

referred to as a `UniversalLocation`.

The `GlobalConsensus` junction refers to a global consensus system, and takes a `NetworkId` that
specifies a particular remote network. A `UniversalLocation` allows for overarching consensus
systems to communicate using this junction. Sub-consensus systems (i.e., a parachain on Polkadot)

> may refer to other sub-consensus systems (a parachain on Kusama) using a combination of

Reviewer 2: ...other *remote* sub-consensus systems...

> `UniversialLocation` and `MultiLocation` declarations.

Reviewer 2: This is superfluous and misleading. Locations may refer to other locations through a `MultiLocation` value - i.e. a relative path between them.

All locations held in consensus are also referencable through an interior location (i.e. a relative location with no parents) from the Universal Location down.

Author: thank you, addressed all outstanding comments!
